hack/generator/pkg/codegen/pipeline_remove_embedded_resources.go

```diff
@@ -0,0 +1,681 @@
+	specPropertiesTypeName, err := extractSpecPropertiesTypeNameOrNil(types, resource)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't extract spec properties")
+	}
+
+	statusPropertiesTypeName, err := extractStatusPropertiesTypeNameOrNil(types, resource)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't extract status properties")
+	}
+
+	return specPropertiesTypeName, statusPropertiesTypeName, nil
+}
+
+// findAllResourcePropertiesTypes finds the "Properties" type for each resource. The result is a astmodel.TypeNameSet containing
+// each resources "Properties" type.
+func findAllResourcePropertiesTypes(types astmodel.Types) (astmodel.TypeNameSet, error) {
+	resources := types.Where(func(def astmodel.TypeDefinition) bool {
+		_, ok := astmodel.AsResourceType(def.Type())
+		return ok
+	})
+
+	var errs []error
+	result := make(astmodel.TypeNameSet)
+
+	// Identify sub-resources and their "properties", associate them with parent resource
+	// Look through parent resource for subresource properties
+	for _, def := range resources {
+		resource, ok := astmodel.AsResourceType(def.Type())
+		if !ok {
+			// Shouldn't be possible to get here
+			panic(fmt.Sprintf("resource was somehow not a resource: %q", def.Name()))
+		}
+
+		specPropertiesTypeName, statusPropertiesTypeName, err := extractSpecAndStatusPropertiesTypeOrNil(types, resource)
+		if err != nil {
+			errs = append(errs, errors.Wrapf(err, "couldn't extract spec/status properties from %q", def.Name()))
+			continue
+		}
+		if specPropertiesTypeName != nil {
+			result = result.Add(*specPropertiesTypeName)
+		}
+		if statusPropertiesTypeName != nil {
+			result = result.Add(*statusPropertiesTypeName)
+		}
+	}
+
+	if len(errs) > 0 {
+		return nil, kerrors.NewAggregate(errs)
+	}
+
+	return result, nil
+}
+
+// findAllResourceStatusTypes finds the astmodel.TypeName's of each resources Status type. If the resource does not have a Status type then
+// that TypeName is not included in the resulting astmodel.TypeNameSet (obviously).
+func findAllResourceStatusTypes(types astmodel.Types) astmodel.TypeNameSet {
+	resources := types.Where(func(def astmodel.TypeDefinition) bool {
+		_, ok := astmodel.AsResourceType(def.Type())
+		return ok
+	})
+
+	result := make(astmodel.TypeNameSet)
+
+	// Identify sub-resources and their "properties", associate them with parent resource
+	// Look through parent resource for subresource properties
+	for _, def := range resources {
+		resource, ok := astmodel.AsResourceType(def.Type())
+		if !ok {
+			// Shouldn't be possible to get here
+			panic(fmt.Sprintf("resource was somehow not a resource: %q", def.Name()))
+		}
+
+		statusName, ok := astmodel.AsTypeName(resource.StatusType())
+		if !ok {
+			continue
+		}
+
+		result = result.Add(statusName)
+	}
+
+	return result
+}
+
+func followTypeNameExtractPropertiesType(types astmodel.Types, typeName astmodel.TypeName) (*astmodel.TypeName, error) {
```

Comment:
Changed to `extractPropertiesType` (agreed name was poor before). Kept nonexported because I'm going to move this into its own module and it's really an implementation detail. Left `typeName` parameter as is though - I think calling it `resource` is confusing because that suggests it's the `TypeName` of the `Resource` which it isn't - it's the `typeName` of a `Spec` or `Status`.

Reply:
Let me know what you think of the above

Reply:
That works. 😀